src/pluggy/callers.py

```diff
@@ -90,13 +90,20 @@ def _wrapped_call(wrap_controller, func):
     try:
         next(wrap_controller)  # first yield
     except StopIteration:
+        wrap_controller.close()
```

Comment:
It isn't necessary to close() after you've received StopIteration (or any other exception) from a call to send() or next(), since that means the generator is already exhausted. (It's harmless though.)

Reply:
Indeed. I'm now starting to wonder if there's any benefit other then triggering the `finally:` in the case where the wrapper was written with >1 `yield` (a mistake by the user).

Reply:
as far as i understood its to ensure the generator cleans up when there was NO stop and its not clear if the generator is actually exhausted

Reply:
Normally this branch is executed if generator finishes before yielding due to some bug in the wrapper. Notice that it still should be a generator and `yield` should be present somewhere in the function body. See `test_multicall_hookwrapper_yield_not_executed` below. So the generator is almost certainly already exhausted and all `finally` clauses have executed already.

I am unsure if `close()` could do something really useful here. I added it with hope that python implementation might be more aggressive trying to release resources of pathological generators

```python
def bad_generator():
    try:
        raise StopIteration()
    finally:
        yield
```

However such code is not valid:
https://docs.python.org/3/library/exceptions.html#StopIteration
> Changed in version 3.7: Enable PEP 479 for all code by default: a `StopIteration` error raised in a generator is transformed into a `RuntimeError`.

If you believe that `close()` is confusing here, I could remove it from this function and from the code handling "modern"-style hook wrappers.

Reply:
At the point where I left this comment, the generator has raised StopIteration. That means it's exhausted. An exception can't escape from a generator if there's anything the generator still might yield. These uses (close() after catching an exception raised out of the generator, including StopIteration) are superfluous.

The PR *also* uses close() to deal with cases where the generator has yielded too many times. That use is valid and useful. Note that close() still might not succeed in closing the generator, if the generator yields in a `finally` block, but that's a bug in the generator. In such cases close() will raise RuntimeError.

Reply:
>At the point where I left this comment, the generator has raised StopIteration. That means it's exhausted.

This is where I was initially ignorant to the fact that `finally:` blocks are indeed triggered on a **single** `yield` generator that has receives `StopIteration`.

>That use is valid and useful. Note that close() still might not succeed in closing the generator, if the generator yields in a finally block, but that's a bug in the generator. In such cases close() will raise RuntimeError.

Yeah I'm thinking catching these cases is a good approach?

@oremanj do you think there's any merit to consideration for the [GC discussion in the introduction of `generator.close()`](https://www.python.org/dev/peps/pep-0342/#new-generator-method-close)? It seems to me that calling `.close()` will avoid any weird errors triggered from `__del__()` in python's future as well.

Reply:
OK, my example of pathological generator was wrong even for python-2. It is possible to yield more values despite attempts to close generator or to send to it another exception but it after some exception is caught from a generator, it is impossible to get more values and it is completely exhausted.